lib/internal/modules/esm/loader.js

```diff
@@ -408,41 +474,96 @@ class ESMLoader {
 
     const namespaces = await PromiseAll(new SafeArrayIterator(jobs));
 
-    return wasArr ?
-      namespaces :
-      namespaces[0];
+    if (!wasArr) return namespaces[0];
+
+    for (let i = 0; i < count; i++) {
+      const namespace = ObjectCreate(null);
+      namespace.url = specifiers[i];
+      namespace.exports = namespaces[i];
+
+      namespaces[i] = namespace;
+    }
+
+    return namespaces;
   }
 
   /**
    * Provide source that is understood by one of Node's translators.
    *
-   * The internals of this WILL change when chaining is implemented,
-   * depending on the resolution/consensus from #36954
-   * @param {string} url The URL/path of the module to be loaded
+   * @param {URL['href']} url The URL/path of the module to be loaded
    * @param {object} context Metadata about the module
-   * @returns {object}
+   * @returns {{ format: ModuleFormat, source: ModuleSource }}
    */
   async load(url, context = {}) {
-    const defaultLoader = this.#loaders[0];
+    const loaders = this.#loaders;
+    let i = loaders.length - 1;
+    let {
+      fn: loader,
+      url: loaderFilePath,
+    } = loaders[i];
+    let chainFinished = i === 0;
+
+    function next(nextUrl, ctx = context) {
```

Comment:
I find it confusing to use function declaration syntax not at the top level; how far up does this get hoisted? If we use assignment syntax we can avoid that question/confusion.

```suggestion
    const next = function(nextUrl, ctx = context) {
```

Reply:
It gets hoisted within its context (eg `ESMLoader::resolve()`, `ESMLoader::load()`). I think there is no danger to mitigate.

The suggestion creates an anonymous function, which is troublesome for stack-tracing.

I think this is not an improvement 😰

Reply:
V8 infers the function name from the variable name:

```shell
node --eval 'const boom = function() { throw new Error() }; boom()'
[eval]:1
const boom = function() { throw new Error() }; boom()
                          ^

Error
    at boom ([eval]:1:33)
    at [eval]:1:48
    at Script.runInThisContext (node:vm:129:12)
    at Object.runInThisContext (node:vm:305:38)
    at node:internal/process/execution:76:19
    at [eval]-wrapper:6:22
    at evalScript (node:internal/process/execution:75:60)
    at node:internal/main/eval_string:27:3

Node.js v18.0.0
```

If you really want to be explicit you can give the anonymous function a name: `const next = function next(nextUrl, ctx = context) {`. Though because of the above, this is redundant.